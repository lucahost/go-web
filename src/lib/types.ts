export type GoBoard = {
    identifier: string
    fields: Field[]
    currentPlayer: Player
    players: [Player, Player]
    history: Field[]
    captures: Field[]
    status: GameState
    pass: boolean
    height: number
    width: number
}

export type User = {
    id: number
    email: string
    name?: string
}

export type Game = {
    id: number
    title: string
    createdAt: Date
    updatedAt: Date
    authorId: number | null
}

export type Player = {
    identifier: string
    name: string
    color: PlayerColor
}

export type Vertex = [number, number]

export type Field = {
    vertex: Vertex
    location: string
    color: PlayerColor
}

export enum GameState {
    INITIALIZED,
    RUNNING,
    ENDED,
    ERROR = 99,
}

export enum PlayerColor {
    WHITE = -1,
    EMPTY = 0,
    BLACK = 1,
}

export enum FieldLocation {
    BLACK_STONE = 'b',
    WHITE_STONE = 'w',
    MARKER = '-',
    MIDDLE = 'm',
    DOWN = 'd',
    DOWN_LEFT = 'dl',
    DOWN_RIGHT = 'dr',
    LEFT = 'l',
    RIGHT = 'r',
    UP = 'u',
    UP_LEFT = 'ul',
    UP_RIGHT = 'ur',
}

export enum HttpMethod {
    GET = 'GET',
    POST = 'POST',
    PATCH = 'PATCH',
    DELETE = 'DELETE',
}
