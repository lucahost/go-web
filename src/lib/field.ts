import { Field, PlayerColor, Vertex } from './types'

export const createField = (
    vertex: Vertex,
    color: PlayerColor = PlayerColor.EMPTY
): Field => ({
    vertex,
    color,
})
