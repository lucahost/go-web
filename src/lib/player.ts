import { Player, PlayerColor } from './types'

export const createPlayer = (name: string, color: PlayerColor): Player => ({
    identifier: 'uuid',
    name,
    color: color,
})
