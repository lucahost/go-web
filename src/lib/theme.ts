import { createGlobalStyle } from 'styled-components'

export const GlobalStyle = createGlobalStyle`
  * {
    box-sizing: border-box;
  }

  html,
  body,
  #__next {
    height: 100%;
    width: 100%;
  }

    body {
   margin: 0;
    padding: 0;
 font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto",
  "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans",
  "Helvetica Neue", sans-serif;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
    color: white;
    background-color: #414246;
  }

  #root {
  display: flex;
  flex-direction: column;
}
`
