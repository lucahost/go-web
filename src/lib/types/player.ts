export enum PlayerColor {
    BLACK = 'Black',
    WHITE = 'White',
}

export interface Player {
    color: PlayerColor
    name: string
}
