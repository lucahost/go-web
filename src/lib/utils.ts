export const chunk = <T>(input: T[], size: number) => {
    return input.reduce((arr: T[][], item, idx) => {
        return idx % size === 0
            ? [...arr, [item]]
            : [...arr.slice(0, -1), [...arr.slice(-1)[0], item]]
    }, [])
}

export const arrayEquals = (a: number[], b: number[]) =>
    a.length === b.length && a.every((v, i) => v === b[i])

export const withoutDuplicates = <T>(list: T[]) =>
    list.filter((obj, index) => {
        const _obj = JSON.stringify(obj)
        return index === list.findIndex(obj => JSON.stringify(obj) === _obj)
    })
