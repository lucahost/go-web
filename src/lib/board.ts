import { Field, FieldLocation, PlayerColor, Vertex } from './types'
import { createField } from './field'

export const generateBoardLayout = (size: number): Field[] => {
    const board = new Array<Field>()

    // Iterate n rows
    for (let row = 1; row <= size; row++) {
        // Iterate n columns
        for (let col = 1; col <= size; col++) {
            if (row === 1 && col === 1) {
                // Up left
                board.push(createField([row, col], FieldLocation.UP_LEFT))
            } else if (row === size && col === 1) {
                // Up right
                board.push(createField([row, col], FieldLocation.DOWN_LEFT))
            } else if (row === 1 && col === size) {
                // Down left
                board.push(createField([row, col], FieldLocation.UP_RIGHT))
            } else if (row === size && col === size) {
                // Down right
                board.push(createField([row, col], FieldLocation.DOWN_RIGHT))
            } else if (row === 1) {
                // Left
                board.push(createField([row, col], FieldLocation.UP))
            } else if (row === size) {
                // Right
                board.push(createField([row, col], FieldLocation.DOWN))
            } else if (col === 1) {
                // Up
                board.push(createField([row, col], FieldLocation.LEFT))
            } else if (col === size) {
                // Right
                board.push(createField([row, col], FieldLocation.RIGHT))
            } else {
                // Markers for 9x9 goban
                if (
                    size === 9 &&
                    ((row === 5 && col === 5) ||
                        (row === 3 && col === 3) ||
                        (row === 3 && col === 7) ||
                        (row === 7 && col === 3) ||
                        (row === 7 && col === 7))
                ) {
                    // Middle marked
                    board.push(createField([row, col], FieldLocation.MARKER))
                } else {
                    // Middle unmarked
                    board.push(createField([row, col], FieldLocation.MIDDLE))
                }
            }
        }
    }
    return board
}

export const withNewFieldColor = (
    board: Field[],
    vertex: Vertex,
    color: PlayerColor
): Field[] => {
    const index = board.findIndex(
        field => field.vertex[0] === vertex[0] && field.vertex[1] === vertex[1]
    )
    board[index] = {
        ...board[index],
        color,
        location:
            color === PlayerColor.BLACK
                ? FieldLocation.BLACK_STONE
                : FieldLocation.WHITE_STONE,
    }
    return board
}
