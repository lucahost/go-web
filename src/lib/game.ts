import {
    Field,
    Game,
    GameState,
    GoBoard,
    Player,
    PlayerColor,
    Vertex,
} from './types'
import { generateBoardLayout, withNewFieldColor } from './board'
import { withoutDuplicates } from './utils'

export const start = (): GoBoard => {
    const width = 9
    const height = 9

    return {
        status: GameState.INITIALIZED,
        captures: [],
        fields: generateBoardLayout(width),
        height,
        currentPlayer: {
            playerColor: PlayerColor.BLACK,
            gameId: 0,
            userId: 1,
        },
        history: [],
        identifier: '',
        pass: false,
        width,
    }
}

export const move = (game: Game, move: Field): GoBoard => {
    let board = game.board as GoBoard
    if (typeof game.board === 'string') {
        board = JSON.parse(game.board) as GoBoard
    }

    if (game.players == undefined) {
        throw 'no players on game'
    }

    if (!isInBounds(board, move)) {
        // Check if move is in bounds
        throw new Error(`Move on location ${move.vertex} is out of bounds`)
    }
    // Check if the color of the move matches the current players color
    if (board?.currentPlayer?.playerColor !== move.color) {
        throw new Error(`current player is not ${move.color}`)
    }

    if (isOccupied(board, move.vertex)) {
        throw new Error(
            `Move on location ${
                move.vertex
            } is not possible. Already occupied. Fields ${board.fields.map(
                f => `${f.vertex}, `
            )}`
        )
    }
    if (isSuicide(board, move.vertex, move.color)) {
        // Handle suicide
        throw new Error('Suicide')
    }

    // Handle Ko
    if (isKo(board, move)) {
        throw new Error()
    }

    // From here on: Valid move !
    // Handle capture
    board = handleCapture(board, move.vertex, move.color)

    // Add move to fields
    board = setStone(board, move)

    // Switch current player
    board = switchPlayer(board, game.players)

    // Reset passes on players if not a double-pass
    board = resetPass(board)
    // Add history
    board = addHistory(board, move)

    // Change state to running
    return { ...board, status: GameState.RUNNING }
}

export const pass = (board: GoBoard): GoBoard => {
    if (board.pass) {
        // TODO: Handle double-pass
        return board
    } else {
        return { ...board, pass: true }
    }
}

export const end = (): void => {
    return
}

// TODO: separate functions below here

export const isInBounds = (board: GoBoard, move: Field): boolean => {
    return (
        move.vertex[0] >= 0 &&
        move.vertex[0] - 1 < board.height &&
        move.vertex[1] >= 0 &&
        move.vertex[1] - 1 < board.width
    )
}

export const isOccupied = (board: GoBoard, vertex: Vertex): boolean => {
    if (board.fields.length == 0) return false

    return findFieldOnBoardByVertex(board, vertex).color !== PlayerColor.EMPTY
}

/*
 * A stone that would have no liberties (or fill the last liberty of your group)
 * must not be placed unless the move captures something. In that case it is
 * allowed (because the capture gives you at least one liberty).
 */
export const isSuicide = (
    board: GoBoard,
    vertex: Vertex,
    playerColor: PlayerColor
): boolean => {
    // Check if the move would capture something
    // To not pass references to the original board around we need to
    // perform a deep copy of the board
    const boardDeepCopy = JSON.parse(JSON.stringify(board)) as GoBoard
    const boardAfterHandleCapture = handleCapture(
        boardDeepCopy,
        vertex,
        playerColor
    )
    if (board.captures < boardAfterHandleCapture.captures) {
        return false
    }
    // Check Liberties on move
    const boardAfterMove = setStone(boardDeepCopy, {
        vertex,
        color: playerColor,
    })
    return getGroupLiberties(boardAfterMove, vertex).length === 0
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isKo = (board: GoBoard, move: Field): boolean => {
    // if board would look the same after performing the move
    // return true
    // else
    return false
}

export const handleCapture = (
    board: GoBoard,
    vertex: Vertex,
    playerColor: PlayerColor
): GoBoard => {
    // get neighbors of move that are of opposite color
    // if the direct neighbor group only has a single liberty
    // which is exactly the current move
    // then move the whole group to captures
    const captures: Field[] = []
    for (const neighbor of getDirectNeighborFieldsOfOppositeColor(
        board,
        vertex,
        playerColor
    )) {
        // get group
        const group = getGroupByVertex(board, neighbor.vertex)
        // check group liberties of neighbors of opposite color
        const groupLiberties = getGroupLiberties(board, neighbor.vertex)
        // if 0 -> remove & add to board.captures
        if (
            groupLiberties.length === 1 &&
            groupLiberties[0].vertex[0] === vertex[0] &&
            groupLiberties[0].vertex[1] === vertex[1]
        ) {
            captures.push(...group)
        }
    }
    return boardWithGroupToCaptures(board, captures)
}

const getOppositeColor = (color: PlayerColor) =>
    color === PlayerColor.BLACK ? PlayerColor.WHITE : PlayerColor.BLACK

const boardWithGroupToCaptures = (board: GoBoard, group: Field[]): GoBoard => {
    for (const field of withoutDuplicates(group)) {
        board.captures.push({
            vertex: field.vertex,
            color: field.color,
        })
        findFieldOnBoardByVertex(board, field.vertex).color = PlayerColor.EMPTY
    }
    return board
}

export const setStone = (board: GoBoard, move: Field): GoBoard => {
    const boardField = findFieldOnBoardByVertex(board, move.vertex)
    boardField.color = move.color

    board.fields = withNewFieldColor(board.fields, move.vertex, move.color)
    return board
}

export const switchPlayer = (board: GoBoard, players: Player[]): GoBoard => {
    if (players?.length !== 2) {
        throw new Error('Incorrect count of players in game')
    }

    const nextPlayer = players?.find(
        p => p.userId !== board?.currentPlayer?.userId
    )

    if (!nextPlayer) {
        throw new Error('Exception loading next player')
    }
    return { ...board, currentPlayer: nextPlayer }
}

export const resetPass = (board: GoBoard): GoBoard => ({
    ...board,
    pass: false,
})

export const addHistory = (board: GoBoard, move: Field): GoBoard => {
    board.history.push(move)
    return board
}

export const findFieldOnBoardByVertex = (
    board: GoBoard,
    vertex: Vertex
): Field => {
    const boardField = board.fields.find(
        field => field.vertex[0] === vertex[0] && field.vertex[1] === vertex[1]
    )

    if (!boardField) {
        throw new Error('Move does not exist in board')
    }

    return boardField
}

export const getLiberties = (board: GoBoard, vertex: Vertex): Field[] => {
    const directNeighborFields = getDirectNeighborFields(board, vertex)

    return directNeighborFields.filter(
        field => field.color === PlayerColor.EMPTY
    )
}

export const getGroupLiberties = (board: GoBoard, vertex: Vertex): Field[] => {
    const liberties: Field[] = []
    const group = getGroupByVertex(board, vertex)

    for (const field of group) {
        liberties.push(...getLiberties(board, field.vertex))
    }

    // only unique fields
    return withoutDuplicates(liberties)
}

export const getGroupByVertex = (
    board: GoBoard,
    vertex: Vertex,
    group: Field[] = []
): Field[] => {
    // find the current field on the board by it's vertex
    const field = findFieldOnBoardByVertex(board, vertex)
    // find the direct neighbors of the same color of the current field
    const directNeighborFields = getDirectNeighborFields(board, vertex).filter(
        neighborField => neighborField.color === field.color
    )
    // only consider yet unknown neighbors and the current field itself
    const newDirectNeighborFields = getNewUniqueFields(group, [
        ...directNeighborFields,
        field,
    ])
    // set the yet known group
    group.push(...newDirectNeighborFields)
    // for each of the new neighbors recursively add any newly identified fields
    for (const neighbor of newDirectNeighborFields) {
        group.push(
            ...getNewUniqueFields(
                group,
                getGroupByVertex(board, neighbor.vertex, group)
            )
        )
    }
    return group
}

const getNewUniqueFields = (fields: Field[], newFields: Field[]): Field[] => {
    const newUniqueFields: Field[] = []
    for (const newField of newFields) {
        if (!fields.includes(newField)) {
            newUniqueFields.push(newField)
        }
    }
    return newUniqueFields
}

const getDirectNeighborFieldsOfOppositeColor = (
    board: GoBoard,
    vertex: Vertex,
    playerColor: PlayerColor
): Field[] =>
    getDirectNeighborFields(board, vertex).filter(
        field => field.color === getOppositeColor(playerColor)
    )

export const getDirectNeighborFields = (
    board: GoBoard,
    vertex: Vertex
): Field[] => {
    const row = vertex[0]
    const col = vertex[1]
    const maxRow = board.height
    const maxCol = board.width

    // Walk rows up if not at max
    return board.fields.filter(field => {
        return (
            // Walk rows up if not out of board
            (row - 1 >= 1 &&
                field.vertex[0] === row - 1 &&
                field.vertex[1] === col) ||
            // Walk rows down if not at max
            (row + 1 <= maxRow &&
                field.vertex[0] === row + 1 &&
                field.vertex[1] === col) ||
            // Walk cols left if not out of board
            (col - 1 >= 1 &&
                field.vertex[1] === col - 1 &&
                field.vertex[0] === row) ||
            // Walk cols right if not at max
            (col + 1 <= maxCol &&
                field.vertex[1] === col + 1 &&
                field.vertex[0] === row)
        )
    })
}
