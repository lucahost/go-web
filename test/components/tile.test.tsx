import { render } from '@testing-library/react'
import Tile from '../../src/components/tile'
import { FieldLocation, PlayerColor, Vertex } from '../../src/lib/types'

describe('Tile', () => {
    it('renders correctly', () => {
        const clickHandler = () => {
            return
        }
        const field = {
            vertex: [0, 0] as Vertex,
            location: FieldLocation.UP_RIGHT,
            color: PlayerColor.EMPTY,
        }
        const { container } = render(
            <Tile
                clickHandler={clickHandler}
                currentPlayer={PlayerColor.BLACK}
                field={field}
            />
        )
        expect(container.firstChild).toMatchSnapshot()
    })
})
