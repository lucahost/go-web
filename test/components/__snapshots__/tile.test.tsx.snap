// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup renders correctly 1`] = `
<img
  class="tile__TileContainer-sc-34hp8n-0 QdBAb"
  src="/Go_ur.svg"
/>
`;

exports[`Tile renders correctly 1`] = `
<img
  class="tile__TileContainer-sc-34hp8n-0 QdBAb"
  src="/Go_ur.svg"
/>
`;
